File: Python/SingletonPattern.py
import threading


class Singleton:
    # Class variable to hold the single instance of the class
    _instance = None

    # Class variable to be used as a lock for thread safety
    _lock = threading.Lock()

    # The __new__ method is called when creating a new instance of the class
    def __new__(cls):
        # Use a lock to ensure thread safety during instance creation
        with cls._lock:
            # Check if an instance already exists
            if not cls._instance:
                # If no instance exists, create a new one
                cls._instance = super(Singleton, cls).__new__(cls)
        # Return the single instance of the class
        return cls._instance


# Example usage in a multithreaded environment
def create_singleton_instance():
    # Create an instance of the Singleton class
    singleton_instance = Singleton()
    # Print a message indicating that an instance has been created
    print(f"Singleton instance created: {singleton_instance}")


# Create a list to store thread objects
threads = []

# Create multiple threads to demonstrate thread safety
for _ in range(5):
    # Create a thread with the target function create_singleton_instance
    thread = threading.Thread(target=create_singleton_instance)
    # Append the thread to the list
    threads.append(thread)

# Start all threads
for thread in threads:
    thread.start()

# Wait for all threads to finish
for thread in threads:
    thread.join()

# Output should show that only one instance is created, even with multiple threads

File: Typescript/SingletonPattern.ts
// You can use a third-party library for async mutex in TypeScript
import { Mutex } from "async-mutex";

class Singleton {
  // Class variable to hold the single instance of the class
  private static _instance: Singleton | null = null;

  // Mutex for thread safety
  private static _mutex = new Mutex();

  // Private constructor to prevent external instantiation
  private constructor() {}

  // Static method to get the single instance of the class
  public static getInstance(): Singleton {
    return this._mutex.runExclusive(async () => {
      if (!this._instance) {
        this._instance = new Singleton();
      }
      return this._instance;
    });
  }

  // Override the toString method to print custom content
  public toString(): string {
    return "I am a Singleton instance!";
  }
}

// Example usage
async function createSingletonInstance() {
  const singletonInstance = await Singleton.getInstance();
  console.log(`Singleton instance created: ${singletonInstance}`);
}

// Create multiple promises to demonstrate asynchronous usage
const promises: Promise<void>[] = [];
for (let i = 0; i < 5; i++) {
  promises.push(createSingletonInstance());
}

// Wait for all promises to finish
Promise.all(promises).then(() => {
  console.log("All promises completed.");
});
